src/table-procedure/src/alter.rs

```diff
@@ -0,0 +1,343 @@
+        // Safety: subprocedure id is always set in this state.
+        let sub_id = self.data.subprocedure_id.unwrap();
+
+        // Query subprocedure state.
+        let Some(sub_state) = ctx.provider.procedure_state(sub_id).await? else {
+            logging::info!(
+                "On engine alter table {}, subprocedure not found, sub_id: {}",
+                self.data.request.table_name,
+                sub_id
+            );
+
+            // If the subprocedure is not found, we create a new subprocedure with the same id.
+            let engine_ctx = EngineContext::default();
+            let procedure = self
+                .engine_procedure
+                .alter_table_procedure(&engine_ctx, self.data.request.clone())
+                .map_err(Error::from_error_ext)?;
+            return Ok(Status::Suspended {
+                subprocedures: vec![ProcedureWithId {
+                    id: sub_id,
+                    procedure,
+                }],
+                persist: true,
+            });
+        };
+
+        match sub_state {
+            ProcedureState::Running | ProcedureState::Retrying { .. } => Ok(Status::Suspended {
+                subprocedures: Vec::new(),
+                persist: false,
+            }),
+            ProcedureState::Done => {
+                logging::info!(
+                    "On engine alter table {}, done, sub_id: {}",
+                    self.data.request.table_name,
+                    sub_id
+                );
+                // The sub procedure is done, we can execute next step.
+                if self.data.request.is_rename_table() {
+                    // We also need to rename the table in the catalog.
+                    self.data.state = AlterTableState::RenameInCatalog;
+                    Ok(Status::executing(true))
+                } else {
+                    // If this isn't a rename operation, we are done.
+                    Ok(Status::Done)
+                }
+            }
+            ProcedureState::Failed { .. } => {
+                // Return error if the subprocedure is failed.
+                SubprocedureFailedSnafu {
+                    subprocedure_id: sub_id,
+                }
+                .fail()?
+            }
+        }
+    }
+
+    async fn on_rename_in_catalog(&mut self) -> Result<Status> {
+        // Safety: table id is available in this state.
+        let table_id = self.data.table_id.unwrap();
+        if let AlterKind::RenameTable { new_table_name } = &self.data.request.alter_kind {
+            let rename_req = RenameTableRequest {
+                catalog: self.data.request.catalog_name.clone(),
+                schema: self.data.request.schema_name.clone(),
+                table_name: self.data.request.table_name.clone(),
+                new_table_name: new_table_name.clone(),
+                table_id,
+            };
+
+            self.catalog_manager
+                .rename_table(rename_req)
```

Comment:
Just find out that if to procedure runs concurrently to rename two different table to the same name, the later rename operation will override previous table with the same name. 

https://github.com/GrepTimeTeam/greptimedb/blob/da66138e809975e1f88b80cd8cb0457de1946e8a/src/catalog/src/local/memory.rs#L325-L337